dhis-2/dhis-web/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/security/JwtBearerTokenTest.java

```diff
@@ -0,0 +1,283 @@
+        assertEquals( "invalid_token", error.getMessage() );
+        assertEquals( "Found no matching DHIS2 user for the mapping claim:'email' with the value:'admin@dhis2.org'",
+            error.getDevMessage() );
+    }
+
+    @Test
+    public void testNoClientMatch()
+    {
+        String providerURI = "testprovidertwo.com";
+
+        setupTestingProvider( "client-2", "testprovidertwo", providerURI );
+
+        String tokenValue = createJwt( providerURI, CLIENT_ID_1, DEFAULT_MAPPING_CLAIM, DEFAULT_EMAIL ).getTokenValue();
+
+        JsonError error = GET( tokenValue, "/me" ).error();
+        assertEquals( HttpStatus.UNAUTHORIZED.value(), error.getHttpStatusCode() );
+        assertEquals( "invalid_token", error.getMessage() );
+        assertEquals( "Invalid audience", error.getDevMessage() );
+    }
+
+    private void setupGoogleProvider( String clientId )
+    {
+        Properties config = new Properties();
+        config.put( ConfigurationKey.OIDC_PROVIDER_GOOGLE_CLIENT_ID.getKey(), clientId );
+        config.put( ConfigurationKey.OIDC_PROVIDER_GOOGLE_CLIENT_SECRET.getKey(), "secret" );
+        DhisOidcClientRegistration parse = GoogleProvider.parse( config );
+        dhisOidcProviderRepository.addRegistration( parse );
+    }
+
+    private void setupTestingProvider( String clientId, String providerName, final String providerURI )
+    {
+        Properties config = new Properties();
+
+        config.put( "oidc.provider." + providerName + ".client_id", clientId );
+        config.put( "oidc.provider." + providerName + ".client_secret", "secret" );
+        config.put( "oidc.provider." + providerName + ".issuer_uri", "https://" + providerURI );
+        config.put( "oidc.provider." + providerName + ".authorization_uri", "https://" + providerURI + "/authorize" );
+        config.put( "oidc.provider." + providerName + ".token_uri", "https://" + providerURI + "/token" );
+        config.put( "oidc.provider." + providerName + ".user_info_uri", "https://" + providerURI + "/userinfo" );
+        config.put( "oidc.provider." + providerName + ".jwk_uri", "https://" + providerURI + "/jwk" );
+
+        GenericOidcProviderConfigParser.parse( config ).forEach( dhisOidcProviderRepository::addRegistration );
+    }
+
+    WebClient.HttpResponse GET( String token, String url, Object... args )
+    {
+        return baseWebRequest( token, get( substitutePlaceholders( url, args ) ), "" );
+    }
+
+    WebClient.HttpResponse baseWebRequest( String token, MockHttpServletRequestBuilder request, String body )
+    {
+        return authWebRequest( token, request );
+    }
+
+    public WebClient.HttpResponse authWebRequest( String token, MockHttpServletRequestBuilder request )
+    {
+        return failOnException(
+            () -> new WebClient.HttpResponse(
+                mvc.perform( request.header( "Authorization", "Bearer " + token ) ).andReturn().getResponse() ) );
+    }
```

Comment:
Here they are - these should be moved to abstract base test and that implements the `AuthentificatedWebClient` interface so this becomes reusable.